go/mcap/new_reader.go

```diff
@@ -0,0 +1,510 @@
+package mcap
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"math"
+	"sort"
+
+	"github.com/klauspost/compress/zstd"
+	"github.com/pierrec/lz4/v4"
+)
+
+type ReadOptions struct {
+	// optional callback which returns whether to include a given channel of messages in the results.
+	ChannelFilter func(schema *Schema, channel *Channel) bool
+	// optional callback which returns whether to include a given attachment in the results.
+	AttachmentFilter func(name string) bool
+	// optional callback which returns whether to include a given metadata record in the results.
+	MetadataFilter func(name string) bool
+	// Messages and Attachments with LogTimes in the range [StartTime, EndTime) will be included.
+	StartTime uint64
+	EndTime   uint64
+
+	// Callbacks to handle parsed records of interest. If a non-nil error is returned,
+	// reading will be aborted.
+	OnMessage    func(*Schema, *Channel, *Message) error
+	OnAttachment func(*Attachment) error
+	OnMetadata   func(*Metadata) error
+	OnStatistics func(*Statistics) error
+}
+
+type CallbackReader struct {
+	opts ReadOptions
+}
+
+func NewCallbackReader(opts ReadOptions) (*CallbackReader, error) {
+	if opts.OnMessage == nil && opts.ChannelFilter != nil {
+		return nil, fmt.Errorf("must supply OnMessage if ChannelFilter is supplied")
+	}
+	if opts.OnAttachment == nil && opts.AttachmentFilter != nil {
+		return nil, fmt.Errorf("must supply OnAttachment if AttachmentFilter is supplied")
+	}
+	if opts.OnMetadata == nil && opts.MetadataFilter != nil {
+		return nil, fmt.Errorf("must supply OnMetadata if MetadataFilter is supplied")
+	}
+	return &CallbackReader{opts: opts}, nil
+}
+
+func (m *ReadOptions) upperTimeBound() uint64 {
+	if m.EndTime == 0 {
+		return math.MaxUint64
+	}
+	return m.EndTime
+}
+
+//
+func (cbr *CallbackReader) Read(reader io.Reader) error {
```

Comment:
it seems more logical to me that CallbackReader would be instantiated on an io.Reader, which would get wrapped in a lexer and attached to the returned CallbackReader, and that the userfacing API would then accept a start/end/topics (and options) rather than a reader.

The current implementation allows me to call Read with the same set of (start/stop/topics) against multiple input files. But I can't think when I would want that.  It seems more useful to me to be able to call Read with different sets of start/stop/topics, against the same input file.